File: implementation/src/search/hmax.py
from heapq import heappush, heappop

def hmax(task, state, operator_costs=None):
    if operator_costs is None:
        operator_costs = {op:op.cost for op in task.operators}
    hmax, closed, precondition_choice_function = {}, {}, {}
    for v in task.variables:
        hmax[v] = float("infinity")
        closed[v] = False
    for op in task.operators:
        op.unsatisfied_preconditions = len(op.precondition)
    # nodes are stored with the key (hmax, depth) to use depth for tie-breaking
    # this gives better (more) landmarks
    heap = []
    for v in state:
        hmax[v] = 0
        heappush(heap, ((0,0),v))

    while heap:
        ((key, depth), u) = heappop(heap)
        if hmax[u] < key:
            # the hmax value of this key was decreased after inserting it
            # this could be done with the min-heap operation decrease-key,
            # but it is faster to insert duplicates and ignore the ones with the old key
            continue
        closed[u] = True
        for op in task.precondition_to_operators[u]:
            if operator_costs[op] == -1:
                continue
            op.unsatisfied_preconditions -= 1
            if op.unsatisfied_preconditions == 0:
                # We discovered all preconditions of this action.
                # Since preconditions are expanded in order of their hmax-value,
                # one of the maximizing preconditions is always expanded last
                # can just choose u as precondition
                # (could also iterate over a.preconditions and search for other preconditions)
                precondition_choice_function[op] = u
                for v in op.effect:
                    if closed[v]:
                        continue
                    successor_cost = hmax[u] + operator_costs[op]
                    if successor_cost < hmax[v]:
                        hmax[v] = successor_cost
                        heappush(heap, ((successor_cost, depth+1), v))
    result = max([hmax[g] for g in task.goal])
    return result, hmax, precondition_choice_function

File: implementation/src/cpp-module/src/hmax.cpp
#include "foreach.h"
#include "hmax.h"
#include "PriorityQueue.h"

int hmax(RelaxedTask &task) {
    VariableSet initialState;
    initialState.add(task.init);
    return hmax(task, initialState);
}

int hmax(RelaxedTask &task, VariableSet &state) {
    OperatorCosts operatorCosts;
    foreach(RelaxedOperator &op, task.operators) {
        operatorCosts[&op] = op.cost;
    }
    return hmax(task, state, operatorCosts);
}

int hmax(RelaxedTask &task, VariableSet &state, OperatorCosts &operatorCosts) {
    // TODO closed list???
    for (unsigned int i = 0; i < task.variables.size(); ++i) {
        task.variables[i].hmax = UNREACHABLE;
        // TODO initialize closed list to false
    }
    for (unsigned int i = 0; i < task.operators.size(); ++i) {
        task.operators[i].unsatisfiedPreconditions = task.operators[i].preconditions.size();
    }
    PriorityQueue queue;
    foreach(Variable *var, state) {
        var->hmax = 0;
        queue.push(var, 0, 0);
    }

    while (!queue.empty()) {
        int hmax, depth;
        Variable *var = queue.pop(hmax, depth);
        if (var == NULL) {
            break;
        }
        // TODO set closed[var] = True
        foreach(RelaxedOperator *op, var->precondition_of) {
            if (operatorCosts[op] == FORBIDDEN) {
                continue;
            }
            op->unsatisfiedPreconditions--;
            if (op->unsatisfiedPreconditions == 0) {
                // We discovered all preconditions of this action.
                // Since preconditions are expanded in order of their hmax-value,
                // one of the maximizing preconditions is always expanded last
                // can just choose var as precondition
                // (could also iterate over op.preconditions and search for other preconditions)
                op->preconditionChoice = var;
                foreach(Variable *effect, op->effects) {
                    // TODO if closed[effect]: continue
                    int successorCost = hmax + operatorCosts[op];
                    if (successorCost < effect->hmax) {
                        queue.decreaseKey(effect, successorCost, depth+1);
                    }
                }
            }
        }
    }
    return task.goal->hmax;
}
